lib/src/config_document_parser.cc

```diff
@@ -0,0 +1,632 @@
+                                                    path *last_path)
+    {
+        string prev_field_name = last_path == nullptr ? "" : last_path->render();
+
+        string part;
+        if (bad_token == tokens::end_token()->to_string()) {
+            // EOF require special handling for the error to make sense
+            if (!prev_field_name.empty()) {
+                part = message + " (if you intended '" + prev_field_name
+                       + "' to be part of a value, instead of a key, try adding double quotes around the whole value";
+            } else {
+                return message;
+            }
+        } else {
+            if (!prev_field_name.empty()) {
+                part = message + " (if you intended " + bad_token
+                       + " to be part of the value for '" + prev_field_name +
+                        "', try enclosing the value in double quotes";
+            } else {
+                part = message + " (if you intended " + bad_token
+                       + " to be part of a key or string value, try enclosing the key or value in double quotes";
+            }
+        }
+
+        if (inside_equals) {
+            return part + ", or you may be able to rename the file .properties rather than .conf)";
+        } else {
+            return part + ")";
+        }
+    }
+
+    shared_node_value parse_context::parse_value(shared_token t) {
+        shared_node_value v = nullptr;
+        int starting_equals_count = _equals_count;
+
+        if (t->get_token_type() == token_type::VALUE || t->get_token_type() == token_type::UNQUOTED_TEXT ||
+                t->get_token_type() == token_type::SUBSTITUTION) {
+            v = make_shared<config_node_simple_value>(t);
+        } else if (t->get_token_type() == token_type::OPEN_CURLY) {
+            v = parse_object(true);
+        } else if (t->get_token_type() == token_type::OPEN_SQUARE) {
+            v = parse_array();
+        } else {
+            throw parse_exception(add_quote_suggestion(t->to_string(),
+                                                   "Expecting a value but got wrong token: " + t->to_string(),
+                                                   _equals_count > 0));
+        }
+
+        if (_equals_count != starting_equals_count) {
+            throw config_exception("Bug in config parser: unbalanced quals count");
+        }
+
+        return v;
+    }
+
+    shared_ptr<config_node_path> parse_context::parse_key(shared_token token) {
+        if (_flavor == config_syntax::JSON) {
+            if (tokens::is_value_with_type(token, config_value_type::STRING)) {
+                single_token_iterator it(token);
+                return make_shared<config_node_path>(path_parser::parse_path_node_expression(it, nullptr));
+            } else {
+                throw parse_exception("\"Expecting close brace } or a field name here, got " + token->to_string());
```

Comment:
The `\"` looks new compared to the original. https://github.com/typesafehub/config/blob/master/config/src/main/java/com/typesafe/config/impl/ConfigDocumentParser.java#L266